usrp_drivers/n200/n200_driver.cpp

```diff
@@ -24,271 +24,339 @@
-    std::cout << "TRANSMIT tr_time_low " << tr_time_low.get_frac_secs() << std::endl;
-    std::cout << "TRANSMIT atten_time_low " << atten_time_low.get_frac_secs() << std::endl;
-    std::cout << "TRANSMIT scope_sync_low " << scope_sync_low.get_frac_secs() << std::endl;
+    auto time_now = usrp_d.get_current_usrp_time();
+    DEBUG_MSG("TRANSMIT time_zero: " << time_zero.get_frac_secs());
+    DEBUG_MSG("TRANSMIT time_now " << time_now.get_frac_secs());
+    DEBUG_MSG("TRANSMIT timetosendsamples(us) " << driver_packet.timetosendsamples());
+    DEBUG_MSG("TRANSMIT time_to_send_pulse " << time_to_send_pulse.get_frac_secs());
+    DEBUG_MSG("TRANSMIT atten_time_high " << atten_time_high.get_frac_secs());
+    DEBUG_MSG("TRANSMIT tr_time_high " << tr_time_high.get_frac_secs());
+    DEBUG_MSG("TRANSMIT pulse_start_time " << pulse_start_time.get_frac_secs());
+    DEBUG_MSG("TRANSMIT tr_time_low " << tr_time_low.get_frac_secs());
+    DEBUG_MSG("TRANSMIT atten_time_low " << atten_time_low.get_frac_secs());
+    DEBUG_MSG("TRANSMIT scope_sync_low " << scope_sync_low.get_frac_secs());
 
     auto md = TXMetadata();
     md.set_has_time_spec(true);
     md.set_time_spec(pulse_start_time);
+    //The USRP tx_metadata start_of_burst and end_of_burst describe start and end of the pulse
+    //samples.
     md.set_start_of_burst(true);
 
     uint64_t num_samps_sent = 0;
 
-    while (num_samps_sent < samples_per_buff) {
+    //This will loop until all samples are sent to the usrp. Send will block until all samples sent
+    //or timed out(too many samples to send within timeout period). Send has a default timing of
+    //0.1 seconds.
+    while (num_samps_sent < samples_per_buff)
+    {
       auto num_samps_to_send = samples_per_buff - num_samps_sent;
-      std::cout << "TRANSMIT Samples to send " << num_samps_to_send << std::endl;
+      DEBUG_MSG("TRANSMIT Samples to send " << num_samps_to_send);
 
-      num_samps_sent = tx_stream->send(
-        samples, num_samps_to_send, md.get_md());
-
-      std::cout << "TRANSMIT Samples sent " << num_samps_sent << std::endl;
+      //Send behaviour can be found in UHD docs
+      num_samps_sent = tx_stream->send( // REVIEW #37 How do we know if it's timed out? Does it do the fragmentation of packets internal to the send(...) if above the max num samps value?
+        samples, num_samps_to_send, md.get_md()); // REVIEW #0 Should we set the timeout to a reasonable value given our timing constraints? 100ms would be way too long to recover this pulse sequence. Maybe we can make a calculation based on how long this pulse is and set the timeout slightly beyond that - or beyond atten_time_low (at which point you should be in this loop for the next pulse) We can also get_max_num_samps to find out if this call will fragment into multiple packets, not sure of the behaviour there, does it return before the timeout??
+                                                  // REPLY lets talk about this.
```

Comment:
So if we set the timeout appropriately so we could recover for the next  pulse sequence, we would need enough time to throw out the driver packets until eob (so we're at the next start of burst) and send a message or flag saying this pulse sequence failed, and send a message to the receiver to say we failed, computation unit needs to know, etc...

An appropriate timeout would be time_zero - time_now - some_threshold
some_threshold would be a small amount of time determined empirically

Reply:
made a todo for this.